### PR TITLE
Feature/hafs rtcases

### LINKED ISSUE #600: Add Regression Test For HAFSv1
There should be at least four set of regression tests for the current suite of HAFSv1 GSI:
1. 3DEnvar with GDAS Ensemble plus FGAT capability.
2. 4Denvar with GDAS Ensemble plus FGAT capability.
3. Self-cycled 3Envar with HAFS Ensemble.
4. Self-cycled 3Envar with GDAS Ensemble plus HAFS Ensemble.
The test case should be on smaller grid with limited number of ensemble members included.

### LINKED ISSUE #647: update the regression test checks
Some of the regression test checks need to be updated or removed to better reflect the results of the regression tests.
such as  maxmem and timing scalability tests.

Need to check and revise the following two files:

regression/regression_test.sh
regression/regression_test_enkf.sh

### PR DESCRIPTION
**Pull request by JingCheng-NOAA**
Add HAFS related regression test into GSI Ctests to solve the issue #600. 

Four set of regression tests for the current suite of HAFSv1 GSI:

3DEnvar with GDAS Ensemble plus FGAT capability -- "hafs_3denvar_glbens".
4Denvar with GDAS Ensemble plus FGAT capability -- "hafs_4denvar_glbens".
3Envar with self-cycled HAFS Ensemble -- "hafs_3denvar_hafens".
3Envar with GDAS Ensemble plus self-cycled HAFS Ensemble -- "hafs_3denvar_hybens".

Fixes #600 
Partially fixes #647 

**Type of change**

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**How Has This Been Tested?**

These changes are tested under Orion through GSI Ctests.
  
**Checklist**

- [x ] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] New and existing tests pass with my changes
- [x] Any dependent changes have been merged and published

**DUE DATE for this PR is 12/4/2023.** If this PR is not merged into `develop` by this date, the PR will be closed and returned to the developer.

### CODE REVIEW

File: regression/regression_var.sh [hunk 1]
```diff
@@ -4,6 +4,10 @@
 #  To run with hybrid ensemble option on, change HYBENS_GLOBAL and/or HYBENS_REGIONAL from "false" to "true".
 #  These are located at the end of this script.
 
+export local_or_default="/work/noaa/hwrf/noscrub/jcheng/gsi"
```
Comment (RussTreadon-NOAA) [inline regression/regression_var.sh]: Quick note:  Please remove lines 7 to 10.   I understand these are for your testing.  We don't want them committed to the authoritative repo.

**Comment by JingCheng-NOAA** [comment]
Fixed!

**Comment by RussTreadon-NOAA** [comment]
Ran hafs ctests on Orion after removing lines 7-1 from [`regression/regression_vars.sh`](https://github.com/NOAA-EMC/GSI/pull/646/files/e64c9c77edf81651c1f2fcacae4dee4fc4fd3ad0#diff-26d77f0e24907aa6e9a08eada075c51f3ac603b88d6b2698f405af2487673610) .

```
Orion-login-3:/work2/noaa/da/rtreadon/git/gsi/pr645/build$ ctest -R hafs_
Test project /work2/noaa/da/rtreadon/git/gsi/pr645/build
    Start  9: hafs_3denvar_glbens
1/4 Test  #9: hafs_3denvar_glbens ..............   Passed  1873.75 sec
    Start 10: hafs_4denvar_glbens
2/4 Test #10: hafs_4denvar_glbens ..............   Passed  2302.39 sec
    Start 11: hafs_3denvar_hafens
3/4 Test #11: hafs_3denvar_hafens ..............***Failed  1758.92 sec
    Start 12: hafs_3denvar_hybens
```

The `hafs_3denvar_hybens` test did not finish because the _loproc_contrl_ job hit the specified 15 minute wall clock time.  The _updat_ jobs finished within 15 minutes
```
Orion-login-3:/work/noaa/stmp/rtreadon/pr645/tmpreg_hafs_3denvar_hybens$ grep wall */stdout
hafs_3denvar_hybens_hiproc_updat/stdout:The total amount of wall time                        = 669.299156
hafs_3denvar_hybens_loproc_updat/stdout:The total amount of wall time                        = 526.743247
```
The slower running _contrl_ job suggests an Orion workload issue since PR #646 does not alter `gsi.x`.   The _updat_ and _contrl_ `gsi.x` are functionally equivalent.

The `hafs_3denvar_hafens` failure is due to the scalability test
```
The case has Failed the scalability test.
The slope for the update (156.030943 seconds per node) is less than that for the control (167.408557 seconds per node).
```
A check of the _updat_ and _contrl_ wall times shows that the _loproc_contrl_ took noticeably longer time than the _loproc_updat_.   
```
hafs_3denvar_hafens_hiproc_contrl/stdout:The total amount of wall time                        = 307.140940
hafs_3denvar_hafens_hiproc_updat/stdout:The total amount of wall time                        = 304.474938
hafs_3denvar_hafens_loproc_contrl/stdout:The total amount of wall time                        = 474.549497
hafs_3denvar_hafens_loproc_updat/stdout:The total amount of wall time                        = 429.299693
```
Since PR #646 does **not** alter source code, this difference is due to the Orion system load.   Hence, this is a non-critical failure.

**Comment by JingCheng-NOAA** [comment]
I've reduced the total number of HAFS related regression tests to 2 after discussed with Shun.

**Comment by ShunLiu-NOAA** [comment]
@yonghuiweng Could you please review this PR and make regression test on WCOSS and other HPCs that HAFS is using. Thank you.

File: regression/CMakeLists.txt [hunk 2]
```diff
@@ -41,6 +41,7 @@ list(APPEND GSI_REG_TEST_NAMES
   global_3dvar global_4dvar global_4denvar
   hwrf_nmm_d2 hwrf_nmm_d3 rtma
```
Comment (hu5970) [inline regression/CMakeLists.txt]: should be able to remove: hwrf_nmm_d2 hwrf_nmm_d3

**Comment by hu5970** [comment]
@JingCheng-NOAA  For verification, you can use the same check as rrfs:

In "GSI/regression/regression_test.sh", you can change the following two lines:
```
351    elif [[ `expr substr $exp1 1 4` = "rrfs" ]]; then
523 elif [[ `expr substr $exp1 1 4` = "rrfs" ]]; then
```

Add "hafs":
```
351    elif [[ `expr substr $exp1 1 4` = "rrfs" ]] || [[ `expr substr $exp1 1 4` = "hafs" ]]; then
523 elif [[ `expr substr $exp1 1 4` = "rrfs" ]] || [[ `expr substr $exp1 1 4` = "hafs" ]]; then
```

Thanks,
Ming

File: regression/CMakeLists.txt (hunk 2 above)
Comment (JingCheng-NOAA) [reply regression/CMakeLists.txt:42]: Will do after Orion comes back later today.

**Comment by RussTreadon-NOAA** [comment]
**Hera ctests**

FYI, Hera `/scratch1/NCEPDEV/da/Russ.Treadon/git/gsi/pr645` contains a working copy of `feature/hafs_rtcases` with the following local modifications
```
        modified:   regression/CMakeLists.txt 
        modified:   regression/multi_regression.sh
        modified:   regression/regression_namelists.sh
        modified:   regression/regression_namelists_db.sh
        modified:   regression/regression_param.sh
        modified:   regression/regression_var.sh
        deleted:    regression/global_3dvar.sh
        deleted:    regression/global_4dvar.sh
        deleted:    regression/hwrf_nmm_d2.sh
        deleted:    regression/hwrf_nmm_d3.sh
```
These changes remove references to the global_3dvar, global_4dvar, hwrf_nmm_d2, and hwrf_nmm_d3 regression tests.   

The data for the hafs tests has been sync'd to Hera.  The remaining 7 ctests were run with the following results.
```
Hera(hfe05):/scratch1/NCEPDEV/da/Russ.Treadon/git/gsi/pr645/build$ ctest -j 7
Test project /scratch1/NCEPDEV/da/Russ.Treadon/git/gsi/pr645/build
    Start 1: global_4denvar
    Start 2: rtma
    Start 3: rrfs_3denvar_glbens
    Start 4: netcdf_fv3_regional
    Start 5: hafs_4denvar_glbens
    Start 6: hafs_3denvar_hybens
    Start 7: global_enkf
1/7 Test #3: rrfs_3denvar_glbens ..............   Passed  550.21 sec
2/7 Test #4: netcdf_fv3_regional ..............   Passed  967.87 sec
3/7 Test #7: global_enkf ......................   Passed  1441.42 sec
4/7 Test #2: rtma .............................   Passed  1632.62 sec
5/7 Test #6: hafs_3denvar_hybens ..............***Failed  1936.57 sec
6/7 Test #1: global_4denvar ...................   Passed  2210.41 sec
7/7 Test #5: hafs_4denvar_glbens ..............***Failed  2240.49 sec

71% tests passed, 2 tests failed out of 7

Total Test time (real) = 2240.51 sec

The following tests FAILED:
          5 - hafs_4denvar_glbens (Failed)
          6 - hafs_3denvar_hybens (Failed)
```

The hafs failures are non-fatal failures.   Both tests failed the maxmem check
```
The memory for hafs_3denvar_hybens_loproc_updat is 2561532 KBs.  This has exceeded maximum allowable hardware memory limit of 2516582 KBs,
resulting in Failure maxmem of the regression test.
```
```
The memory for hafs_4denvar_glbens_loproc_updat is 2878960 KBs.  This has exceeded maximum allowable hardware memory limit of 2516582 KBs,
resulting in Failure maxmem of the regression test.
```
This check is misleading.  The threshold is set to the per node memory divided by the number of cores.  For Hera this is 96 Gb  divided by 40 cores (2516582 Kb or 2.4 Gb).  The hafs tests run with 20 tasks per Hera node.  As such, neither hafs tests comes close to exceeding the 96 Gb limit for Hera nodes.

As @hu5970 mentioned, some of the current regression test checks should either be removed or revised.   The maxmem and timing scalability tests can probably be removed.

**Comment by JingCheng-NOAA** [comment]
> FYI, Hera `/scratch1/NCEPDEV/da/Russ.Treadon/git/gsi/pr645` contains a working copy of `feature/hafs_rtcases` with the following local modifications
> 
> ```
>         modified:   regression/CMakeLists.txt 
>         modified:   regression/multi_regression.sh
>         modified:   regression/regression_namelists.sh
>         modified:   regression/regression_namelists_db.sh
>         modified:   regression/regression_param.sh
>         modified:   regression/regression_var.sh
>         deleted:    regression/global_3dvar.sh
>         deleted:    regression/global_4dvar.sh
>         deleted:    regression/hwrf_nmm_d2.sh
>         deleted:    regression/hwrf_nmm_d3.sh
> ```
> 
> These changes remove references to the global_3dvar, global_4dvar, hwrf_nmm_d2, and hwrf_nmm_d3 regression tests.
> 
> The data for the hafs tests has been sync'd to Hera. The remaining 7 ctests were run with the following results.
> 
> ```
> Hera(hfe05):/scratch1/NCEPDEV/da/Russ.Treadon/git/gsi/pr645/build$ ctest -j 7
> Test project /scratch1/NCEPDEV/da/Russ.Treadon/git/gsi/pr645/build
>     Start 1: global_4denvar
>     Start 2: rtma
>     Start 3: rrfs_3denvar_glbens
>     Start 4: netcdf_fv3_regional
>     Start 5: hafs_4denvar_glbens
>     Start 6: hafs_3denvar_hybens
>     Start 7: global_enkf
> 1/7 Test #3: rrfs_3denvar_glbens ..............   Passed  550.21 sec
> 2/7 Test #4: netcdf_fv3_regional ..............   Passed  967.87 sec
> 3/7 Test #7: global_enkf ......................   Passed  1441.42 sec
> 4/7 Test #2: rtma .............................   Passed  1632.62 sec
> 5/7 Test #6: hafs_3denvar_hybens ..............***Failed  1936.57 sec
> 6/7 Test #1: global_4denvar ...................   Passed  2210.41 sec
> 7/7 Test #5: hafs_4denvar_glbens ..............***Failed  2240.49 sec
> 
> 71% tests passed, 2 tests failed out of 7
> 
> Total Test time (real) = 2240.51 sec
> 
> The following tests FAILED:
>           5 - hafs_4denvar_glbens (Failed)
>           6 - hafs_3denvar_hybens (Failed)
> ```
> 
> The hafs failures are non-fatal failures. Both tests failed the maxmem check
> 
> ```
> The memory for hafs_3denvar_hybens_loproc_updat is 2561532 KBs.  This has exceeded maximum allowable hardware memory limit of 2516582 KBs,
> resulting in Failure maxmem of the regression test.
> ```
> 
> ```
> The memory for hafs_4denvar_glbens_loproc_updat is 2878960 KBs.  This has exceeded maximum allowable hardware memory limit of 2516582 KBs,
> resulting in Failure maxmem of the regression test.
> ```
> 
> This check is misleading. The threshold is set to the per node memory divided by the number of cores. For Hera this is 96 Gb divided by 40 cores (2516582 Kb or 2.4 Gb). The hafs tests run with 20 tasks per Hera node. As such, neither hafs tests comes close to exceeding the 96 Gb limit for Hera nodes.
> 
> As @hu5970 mentioned, some of the current regression test checks should either be removed or revised. The maxmem and timing scalability tests can probably be removed.

Previously I only tested it on Orion. So the settings for Hera may be off. Please feel free to change it or I can make changes as well.

File: regression/CMakeLists.txt (hunk 2 above)
Comment (JingCheng-NOAA) [reply regression/CMakeLists.txt]: I will remove them and change the verification part as well.

**Comment by RussTreadon-NOAA** [comment]
**Hera ctests (continued)**

Remove maxmem and timing scalability tests from `regression_test.sh` and `regression_test_enkf.sh` in the above mentioned Hera working copy of `hafs-community:feature/hafs_rtcases`.   Rerun seven ctests.  
```
Hera(hfe03):/scratch1/NCEPDEV/da/Russ.Treadon/git/gsi/pr645/build$ ctest -j 7
Test project /scratch1/NCEPDEV/da/Russ.Treadon/git/gsi/pr645/build
    Start 1: global_4denvar
    Start 2: rtma
    Start 7: global_enkf
    Start 4: netcdf_fv3_regional
    Start 3: rrfs_3denvar_glbens
    Start 5: hafs_4denvar_glbens
    Start 6: hafs_3denvar_hybens
1/7 Test #4: netcdf_fv3_regional ..............   Passed  491.83 sec
2/7 Test #3: rrfs_3denvar_glbens ..............   Passed  675.13 sec
3/7 Test #7: global_enkf ......................   Passed  949.82 sec
4/7 Test #6: hafs_3denvar_hybens ..............   Passed  2182.70 sec
5/7 Test #5: hafs_4denvar_glbens ..............   Passed  2308.16 sec
6/7 Test #2: rtma .............................   Passed  2474.11 sec
7/7 Test #1: global_4denvar ...................   Passed  3053.83 sec

100% tests passed, 0 tests failed out of 7

Total Test time (real) = 3053.84 sec
```
All tests pass.  This is an expected result since this PR does not alter code, namelists, modules, or scripts..   This test is relevant to GSI issue #647 and is cross referenced there.

**Comment by hu5970** [comment]
Seems Hera and Orion are good for testing. Is any one working on WCOSS2?

**Comment by ShunLiu-NOAA** [comment]
As suggested on Monday's meeting, someone from HAFS DA might run the test on WCOSS to get familiar with HAFS DA regression test. Yonghui might make a regression test on WCOSS once he gets chance.

**Comment by RussTreadon-NOAA** [comment]
> Seems Hera and Orion are good for testing. Is any one working on WCOSS2?

HAFS data needs to be rsync'd to WCOSS2.

**Comment by yonghuiweng** [comment]
Yes, I’m working on dogwood.

 �

**Comment by RussTreadon-NOAA** [comment]
**WCOSS2 (Dogwood) tests**
Do the following
 - Rsync Hera working copy of `hafs_rtcases` to Dogwood `/lfs/h2/emc/da/noscrub/russ.treadon/git/gsi/pr645`
 - Update with current head of `feature/hafs_rtcases`
 - Rsync hafs regression test data to Dogwood
 - Build `feature/hafs_rtcases`
 - Run seven ctests.  As shown below, all tests pass.  This is an expected result.
```
russ.treadon@dlogin07:/lfs/h2/emc/da/noscrub/russ.treadon/git/gsi/pr645/build> ctest -j 7
Test project /lfs/h2/emc/da/noscrub/russ.treadon/git/gsi/pr645/build
    Start 1: global_4denvar
    Start 2: rtma
    Start 3: rrfs_3denvar_glbens
    Start 4: netcdf_fv3_regional
    Start 5: hafs_4denvar_glbens
    Start 6: hafs_3denvar_hybens
    Start 7: global_enkf
1/7 Test #4: netcdf_fv3_regional ..............   Passed  482.59 sec
2/7 Test #7: global_enkf ......................   Passed  610.86 sec
3/7 Test #3: rrfs_3denvar_glbens ..............   Passed  666.14 sec
4/7 Test #2: rtma .............................   Passed  1088.11 sec
5/7 Test #1: global_4denvar ...................   Passed  1381.85 sec
6/7 Test #6: hafs_3denvar_hybens ..............   Passed  1448.17 sec
7/7 Test #5: hafs_4denvar_glbens ..............   Passed  1630.54 sec

100% tests passed, 0 tests failed out of 7

Total Test time (real) = 1630.56 sec
```
As a reminder, my modified working copy of `feature/hafs_rtcases` removes the maxmem and timing scalability tests.

The two hafs tests have the longest run time in the ctest suite.   Notice that the hafs tests run two outer loops with each inner loop containing 50 iterations.  _Can we reduce the number of inner iterations?_   Reducing the inner loop iteration count is an easy way to reduce wall time.

The global_4denvar test runs 2 outer loops with 5 iterations on the first outer loop and 10 on the second outer loop.  The global_4denvar test could reduce the second outer loop iterations to 5.

We should check the number of inner loop iterations used in other ctests.

**Comment by JingCheng-NOAA** [comment]
I have no issue to reduce the iterations.JingOn Oct 25, 2023, at 21:26, RussTreadon-NOAA ***@***.***> wrote:﻿
WCOSS2 (Dogwood) tests
Do the following

Rsync Hera working copy of hafs_rtcases to Dogwood /lfs/h2/emc/da/noscrub/russ.treadon/git/gsi/pr645
Update with current head of feature/hafs_rtcases
Rsync hafs regression test data to Dogwood
Build feature/hafs_rtcases
Run seven ctests.  As shown below, all tests pass.  This is an expected result.

***@***.***:/lfs/h2/emc/da/noscrub/russ.treadon/git/gsi/pr645/build> ctest -j 7
Test project /lfs/h2/emc/da/noscrub/russ.treadon/git/gsi/pr645/build
    Start 1: global_4denvar
    Start 2: rtma
    Start 3: rrfs_3denvar_glbens
    Start 4: netcdf_fv3_regional
    Start 5: hafs_4denvar_glbens
    Start 6: hafs_3denvar_hybens
    Start 7: global_enkf
1/7 Test #4: netcdf_fv3_regional ..............   Passed  482.59 sec
2/7 Test #7: global_enkf ......................   Passed  610.86 sec
3/7 Test #3: rrfs_3denvar_glbens ..............   Passed  666.14 sec
4/7 Test #2: rtma .............................   Passed  1088.11 sec
5/7 Test #1: global_4denvar ...................   Passed  1381.85 sec
6/7 Test #6: hafs_3denvar_hybens ..............   Passed  1448.17 sec
7/7 Test #5: hafs_4denvar_glbens ..............   Passed  1630.54 sec

100% tests passed, 0 tests failed out of 7

Total Test time (real) = 1630.56 sec

As a reminder, my modified working copy of feature/hafs_rtcases removes the maxmem and timing scalability tests.
The two hafs tests have the longest run time in the ctest suite.   Notice that the hafs tests run two outer loops with each inner loop containing 50 iterations.  Can we reduce the number of inner iterations?   Reducing the inner loop iteration count is an easy way to reduce wall time.
The global_4denvar test runs 2 outer loops with 5 iterations on the first outer loop and 10 on the second outer loop.  The global_4denvar test could reduce the second outer loop iterations to 5.
We should check the number of inner loop iterations used in other ctests.

**Comment by RussTreadon-NOAA** [comment]
**Orion ctests**

Repeat on Orion the steps followed on Hera and WCOSS2.  All seven ctests pass
```
Orion-login-4:/work2/noaa/da/rtreadon/git/gsi/pr645/build$ ctest -j 7
Test project /work2/noaa/da/rtreadon/git/gsi/pr645/build
    Start 1: global_4denvar
    Start 2: rtma
    Start 3: rrfs_3denvar_glbens
    Start 4: netcdf_fv3_regional
    Start 5: hafs_4denvar_glbens
    Start 6: hafs_3denvar_hybens
    Start 7: global_enkf
1/7 Test #4: netcdf_fv3_regional ..............   Passed  542.43 sec
2/7 Test #7: global_enkf ......................   Passed  547.92 sec
3/7 Test #3: rrfs_3denvar_glbens ..............   Passed  604.99 sec
4/7 Test #2: rtma .............................   Passed  1208.90 sec
5/7 Test #1: global_4denvar ...................   Passed  1561.90 sec
6/7 Test #6: hafs_3denvar_hybens ..............   Passed  1873.67 sec
7/7 Test #5: hafs_4denvar_glbens ..............   Passed  2114.52 sec

100% tests passed, 0 tests failed out of 7

Total Test time (real) = 2114.54 sec
```

**Comment by RussTreadon-NOAA** [comment]
Thank you @JingCheng-NOAA for being open to reducing the number of inner loop iterations in the hafs tests.  I'll reduce the inner loop iterations in other tests and rerun the ctests to quantify the run time impact.

**Comment by RussTreadon-NOAA** [comment]
**Reduced inner loop iterations**

Reduce the number of inner loop iterations for several tests.  Rerun ctests on Orion, Hera, and WCOSS2 (Dogwood) with the following results.

_Orion_
```
Orion-login-4:/work2/noaa/da/rtreadon/git/gsi/pr645/build$ ctest -j 7
Test project /work2/noaa/da/rtreadon/git/gsi/pr645/build
    Start 5: hafs_4denvar_glbens
    Start 6: hafs_3denvar_hybens
    Start 1: global_4denvar
    Start 2: rtma
    Start 3: rrfs_3denvar_glbens
    Start 7: global_enkf
    Start 4: netcdf_fv3_regional
1/7 Test #4: netcdf_fv3_regional ..............   Passed  482.46 sec
2/7 Test #7: global_enkf ......................   Passed  488.62 sec
3/7 Test #3: rrfs_3denvar_glbens ..............   Passed  545.26 sec
4/7 Test #2: rtma .............................   Passed  969.11 sec
5/7 Test #6: hafs_3denvar_hybens ..............   Passed  1273.80 sec
6/7 Test #5: hafs_4denvar_glbens ..............   Passed  1452.80 sec
7/7 Test #1: global_4denvar ...................   Passed  1562.07 sec

100% tests passed, 0 tests failed out of 7

Total Test time (real) = 1562.08 sec
```
The previous round of ctests on _Orion_ took `2114.54 sec`.

_Hera_
```
Hera(hfe10):/scratch1/NCEPDEV/da/Russ.Treadon/git/gsi/pr645/build$ ctest -j 7
Test project /scratch1/NCEPDEV/da/Russ.Treadon/git/gsi/pr645/build
    Start 1: global_4denvar
    Start 5: hafs_4denvar_glbens
    Start 6: hafs_3denvar_hybens
    Start 2: rtma
    Start 7: global_enkf
    Start 4: netcdf_fv3_regional
    Start 3: rrfs_3denvar_glbens
1/7 Test #4: netcdf_fv3_regional ..............   Passed  676.55 sec
2/7 Test #3: rrfs_3denvar_glbens ..............   Passed  733.49 sec
3/7 Test #7: global_enkf ......................   Passed  985.35 sec
4/7 Test #2: rtma .............................   Passed  1218.54 sec
5/7 Test #6: hafs_3denvar_hybens ..............   Passed  1406.38 sec
6/7 Test #5: hafs_4denvar_glbens ..............   Passed  1464.25 sec
7/7 Test #1: global_4denvar ...................   Passed  2160.74 sec

100% tests passed, 0 tests failed out of 7

Total Test time (real) = 2160.75 sec
```
The previous round of ctests on _Hera_ took `3053.84 sec`.

_WCOSS2 (Dogwood)_
```
russ.treadon@dlogin08:/lfs/h2/emc/da/noscrub/russ.treadon/git/gsi/pr645/build> ctest -j 7
Test project /lfs/h2/emc/da/noscrub/russ.treadon/git/gsi/pr645/build
    Start 1: global_4denvar
    Start 5: hafs_4denvar_glbens
    Start 6: hafs_3denvar_hybens
    Start 2: rtma
    Start 7: global_enkf
    Start 3: rrfs_3denvar_glbens
    Start 4: netcdf_fv3_regional
1/7 Test #4: netcdf_fv3_regional ..............   Passed  482.63 sec
2/7 Test #3: rrfs_3denvar_glbens ..............   Passed  485.01 sec
3/7 Test #7: global_enkf ......................   Passed  609.06 sec
4/7 Test #2: rtma .............................   Passed  968.42 sec
5/7 Test #6: hafs_3denvar_hybens ..............   Passed  1208.72 sec
6/7 Test #5: hafs_4denvar_glbens ..............   Passed  1209.08 sec
7/7 Test #1: global_4denvar ...................   Passed  1322.42 sec

100% tests passed, 0 tests failed out of 7

Total Test time (real) = 1322.45 sec
```
The previous round of ctests on Dogwood took `1630.56 sec`.

The modified working directory for `hafs_rtcases` including changes in the number of inner loop iterations is found in the following locations:
- Orion:  `/work2/noaa/da/rtreadon/git/gsi/pr645`
- Hera:  `/scratch1/NCEPDEV/da/Russ.Treadon/git/gsi/pr645`
- Dogwood:  `/lfs/h2/emc/da/noscrub/russ.treadon/git/gsi/pr645`

**Comment by JingCheng-NOAA** [comment]
Thanks for the update and the test results look great.

**Comment by RussTreadon-NOAA** [comment]
I updated a working copy of `feature/hafs_rtcases` on Orion in `/work2/noaa/da/rtreadon/git/gsi/update` with my changes.  I committed the changes to my local working copy.  

```
Orion-login-4:/work2/noaa/da/rtreadon/git/gsi/update$  git log | head
commit 1faf99ab9c40ad40403bda7b0ba0c3eed9457be9
Author: RussTreadon-NOAA <russ.treadon@noaa.gov>
Date:   Thu Oct 26 13:35:54 2023 +0000

    clean up, optimize, and simplify regression tests (#600, #647)
```

I can not push the commit to `hafs-community:feature/hafs_rtcases`.   I do not have write permission to this repository.

If the changes in Orion `work2/noaa/da/rtreadon/git/gsi/update` are acceptable to the team, would someone with write permission please commit these changes to `hafs-community:feature/hafs_rtcases`.    

Thank you.

**Comment by JingCheng-NOAA** [comment]
> I updated a working copy of `feature/hafs_rtcases` on Orion in `/work2/noaa/da/rtreadon/git/gsi/update` with my changes. I committed the changes to my local working copy.
> 
> ```
> Orion-login-4:/work2/noaa/da/rtreadon/git/gsi/update$  git log | head
> commit 1faf99ab9c40ad40403bda7b0ba0c3eed9457be9
> Author: RussTreadon-NOAA <russ.treadon@noaa.gov>
> Date:   Thu Oct 26 13:35:54 2023 +0000
> 
>     clean up, optimize, and simplify regression tests (#600, #647)
> ```
> 
> I can not push the commit to `hafs-community:feature/hafs_rtcases`. I do not have write permission to this repository.
> 
> If the changes in Orion `work2/noaa/da/rtreadon/git/gsi/update` are acceptable to the team, would someone with write permission please commit these changes to `hafs-community:feature/hafs_rtcases`.
> 
> Thank you.

Hi Russ,
I will work on  commit these changes to 'hafs-community:feature/hafs_rtcases'. But right now I don't have permission to access the directory: work2/noaa/da/rtreadon/git/gsi/update. Could you please grant me permission to access that folder?
Jing

**Comment by RussTreadon-NOAA** [comment]
@JingCheng-NOAA , please try again.   I executed a `chmod 755` on the directory.

**Comment by JingCheng-NOAA** [comment]
Thanks, it worked.

**Comment by JingCheng-NOAA** [comment]
> I updated a working copy of `feature/hafs_rtcases` on Orion in `/work2/noaa/da/rtreadon/git/gsi/update` with my changes. I committed the changes to my local working copy.
> 
> ```
> Orion-login-4:/work2/noaa/da/rtreadon/git/gsi/update$  git log | head
> commit 1faf99ab9c40ad40403bda7b0ba0c3eed9457be9
> Author: RussTreadon-NOAA <russ.treadon@noaa.gov>
> Date:   Thu Oct 26 13:35:54 2023 +0000
> 
>     clean up, optimize, and simplify regression tests (#600, #647)
> ```
> 
> I can not push the commit to `hafs-community:feature/hafs_rtcases`. I do not have write permission to this repository.
> 
> If the changes in Orion `work2/noaa/da/rtreadon/git/gsi/update` are acceptable to the team, would someone with write permission please commit these changes to `hafs-community:feature/hafs_rtcases`.
> 
> Thank you.

Hi Russ, I know you've removed the the maxmem and timing scalability tests in "regression_test_enkf.sh" and "regression_test.sh" for test purpose. I shouldn't include the changes in these two scripts, correct?

**Comment by RussTreadon-NOAA** [comment]
@JingCheng-NOAA , after talking with @ShunLiu-NOAA it's my impression that we want to bring _all_ my modifications into your branch.   This incudes changes to `regression_test.sh` and `regression_test_enkf.sh`.

**Comment by JingCheng-NOAA** [comment]
Thanks for the clarification. I will include all the changes then.

**Review by ShunLiu-NOAA — DISMISSED** [review]

File: regression/multi_regression.sh [hunk 3]
```diff
@@ -1,22 +1,18 @@
 #!/bin/sh --login
 
-regtests_all="global_3dvar
-              global_4dvar
-              global_4denvar
+regtests_all="global_4dendvar
```
Comment (RussTreadon-NOAA) [inline regression/multi_regression.sh]: `global_4dendvar` should read `global_4denvar`.   Please remove the `d` between `4den` and `var`.

**Comment by ShunLiu-NOAA** [comment]
Thank you @RussTreadon-NOAA and @JingCheng-NOAA. 
Jing, could you please run regression test with all modifications in your branch?
@hu5970 For this PR, is there anything more we still should test?

File: regression/multi_regression.sh (hunk 3 above)
Comment (RussTreadon-NOAA) [reply regression/multi_regression.sh]: Corrected at [549497b](https://github.com/NOAA-EMC/GSI/pull/646/commits/549497b7c9648c77c6897f41ff7b92e7af613fa7).

**Comment by RussTreadon-NOAA** [comment]
@JingCheng-NOAA , we should `git rm` the following files
```
regression/global_3dvar.sh
regression/global_4dvar.sh
```
in `hafs-community:feature/hafs_rtcases`.   `global_4denvar` is the only global gsi test executed by ctest.

**Comment by JingCheng-NOAA** [comment]
Agree. I removed them.

**Review by RussTreadon-NOAA — APPROVED** [review]
Rerun ctests on Hera, Orion, and Dogwood.  All 7 tests pass on each machine.

Approve pending input from others.  
- @hu5970 , do you have any regression test changes you want to commit to `hafs-community:feature/hafs_rtcases`?
- @JingCheng-NOAA , do you have additional changes you would like to make to your branch?

**Comment by JingCheng-NOAA** [comment]
I have no more changes to add.
BTW, I've also tested on ORION and all 7 tests passed with no issues.

Test project /work/noaa/hwrf/save/jcheng/GSIversions/GSI/build
    Start 5: hafs_4denvar_glbens
    Start 1: global_4denvar
    Start 6: hafs_3denvar_hybens
    Start 2: rtma
    Start 3: rrfs_3denvar_glbens
    Start 7: global_enkf
    Start 4: netcdf_fv3_regional
1/7 Test #4: netcdf_fv3_regional ..............   Passed  242.96 sec
2/7 Test #7: global_enkf ......................   Passed  249.75 sec
3/7 Test #3: rrfs_3denvar_glbens ..............   Passed  546.38 sec
4/7 Test #2: rtma .............................   Passed  970.55 sec
5/7 Test #6: hafs_3denvar_hybens ..............   Passed  1338.81 sec
6/7 Test #5: hafs_4denvar_glbens ..............   Passed  1456.55 sec
7/7 Test #1: global_4denvar ...................   Passed  1562.20 sec

100% tests passed, 0 tests failed out of 7

Total Test time (real) = 1562.21 sec

**Pull request merged**